src/solve2!.jl

```diff
@@ -0,0 +1,80 @@
+function create_context(::Problem{T}, optimizer) where {T}
+    model = MOIB.full_bridge_optimizer(MOIU.CachingOptimizer(MOIU.UniversalFallback(MOIU.Model{T}()),
+                                                             optimizer), T)
+
+    return (var_id_to_moi_indices=OrderedDict{UInt64,Vector{MOI.VariableIndex}}(),
+            model=model, T=T, id_to_variables=OrderedDict{UInt64,AbstractVariable}())
+end
+
+
+function add_variables!(model, var::AbstractVariable)
+    var.id_hash == objectid(:constant) && error("Internal error: constant used as variable")
+    return if sign(var) == ComplexSign()
+        error("complex not implemented")
+    else
+        return MOI.add_variables(model, length(var))
+    end
+end
+
+# this is kind of awful; currently needed since sometimes Convex treats numbers as vectors or matrices
+# need to think of a better way, possibly via breaking changes to Convex's syntax
+function promote_size(values)
+    d = only(unique(MOI.output_dimension(v)
+                for v in values
+                if v isa MOI.AbstractFunction || v isa VectorAffineFunctionAsMatrix))
+    return (v isa Number ? fill(v, d) : v for v in values)
+end
+
+scalar_fn(x) = only(MOIU.scalarize(x))
+scalar_fn(v::MOI.AbstractScalarFunction) = v
```

Comment:
I'm not sure this should really be needed. It came about because I ended up passing things that were nomially vectors to the objective function, or adding scalars and vectors. I think part of the problem is that Convex.jl does not distinguish between scalars and vectors (or matrices!); e.g. atoms are not parametric types. This is an inherent source of type instability and means that we can't use dispatch always to route things correctly (i.e. one might imagine that the objective function would always end up as a scalar function just by the layout of the problem, but that doesn't happen currently).